### DISCUSSION TITLE
Metadrive simulator

### DISCUSSION THREAD

**Discussion by VishalP227** [metadriverse/metadrive#440] [Q&A] [answered]
I had a few questions about the possible functionalities of the Metadrive simulator as given below:
1. Is it possible to set the camera to first person view during driving without using the third person view at all ? If so how could we do this ?  Will camera_dist/height in Visualisation and Rendering Config be helpful in doing this ?
2. In the documentation there is information about how to control the ego vehicle, is there a possibility to have a multi player setting where more than one person are controlling two or more ego vehicles in the simulation ?
3. Is there a possibility to run the simulation in a web browser ?

**Comment by QuanyiLi** [comment, marked as answer]
Hi, 

Welcome to MetaDrive and thanks for your interest!

#### Q1: FPP view
You can set the pose and position of the default camera via the environment config. I believe these options can satisfy most requirements. We are going to expose an API taking extrinsic/intrinsic matrix soon.
```
"camera_dist": -0.5, # set -0.5 to move it to the front of the Car
"camera_height": 0.5, 
"camera_pitch": np.pi / 3, # make it look at the ground
"camera_fov": 60,
```

#### Q2: Multi-player Vehicle Control
Nice idea. I believe it can be supported easily. In MetaDrive, all vehicles and objects are controlled by policies, which are inherited from ```class BasePolicy```. Take ```ManulaControlPolicy``` as an example: https://github.com/metadriverse/metadrive/blob/4d77e619b2bc53397c6836e26a17dfc43b5d8cd8/metadrive/policy/manual_control_policy.py#L45
The ```def act``` function decides the steering and throttle for the controlled vehicle by reading the keyboard or joystick input and will be called each frame automatically. Therefore, you can write your own policy to determine how to get the throttle and steering command for each vehicle. This policy can read the input from another set of keys, another device, like the steering wheel, or even another machine (add some communication mechanism). The only thing that needs to care about is binding the target vehicles and different policy instances via ```engine.add_policy(object, policy)```.

#### Q3: Run in web-browser
Well, I am a noob about web services, so I can not give you a certain answer. But I remember our game engine supports web-gl as the rendering backend, so it might be possible. I am not 100% sure about this, sorry. 
Yes, I checked a thread: https://discourse.panda3d.org/t/how-to-run-panda3d-in-browser/27579/2 it is possible, but we have not tried yet.

**Reply by VishalP227** [reply]
Thank you so much for providing the answers. 

I had one another question:
Q4: Is Metadrive simulator compatible with ROS (Robot Operating System) ?

**Reply by QuanyiLi** [reply]
We are working on a ROS bridge, and hope it can be released in one or two months : )
